sql: rollback disallowing statements following schema change

vivekmenezes opened the pull request:
rolls back #15339 #14368 for the most part and #14619 partially.

We used a very big hammer to prevent some statements from
following a schema change. This was done to reduce surprises
but it ended up introducing more surprises. ORMs and schema
migration tools depends on multiple schema changes executing
in the same transaction.

fixes #15269

vivekmenezes committed:
sql: rollback disallowing statements following schema change

rolls back #15339 #14368 for the most part and #14619 partially.

We used a very big hammer to prevent some statements from
following a schema change. This was done to reduce surprises
but it ended up introducing more surprises. ORMs and schema
migration tools depend on multiple schema changes being
staged in the same transaction.

fixes #15269

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15511)
<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Given that we're reverting something that seemed like a useful change, I think we should document somewhere in the code why we need to support the current behavior (i.e. because of ORM compatibility). A quick scan of the code didn't reveal a great place. Perhaps a comment associated with one of the tests being re-added.

Also, my pedantic side notes that this change doesn't allow executing multiple schema changes in a transaction (as noted in the commit message), only "staging" multiple schema changes. 

Lastly, the way we handle schema changes is different than Postgres. We should review the documentation in this area and make sure the differences are clear.

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15511#-:-Kiu2EA_k3QM41bkGOmX:bpgjx1o)*
<!-- Sent from Reviewable.io -->

vivekmenezes committed:
sql: added a test case for multiple schema changes in a txn

the test also includes an insert into an auxiliary table in the
same transaction as a way to record the schema change transaction.

fixes #15297

vivekmenezes commented:
I've changed the commit to say "staging" multiple schema changes and have also added a test
case for the multiple schema change transaction including running an INSERT on an auxiliary table
in the same transaction. which covers #15297

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15511#-:-KivVv8_MzxRb_Ne84cE:bwjcskh)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/sql/schema_changer_test.go, line 1919 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15511#-KivYT2EN2vCree8yFle:-KivYT2EN2vCree8yFlf:b4z5qde) ([raw file](https://github.com/cockroachdb/cockroach/blob/554bca17f681b73cbcf8c25291c878b2b4e8d9b8/pkg/sql/schema_changer_test.go#L1919)):*
> ```Go
> 			`INSERT INTO t.origin VALUES ('c', 'd')`, ``},
> 		// Support multiple schema changes for ORMs: #15269
> 		// Support insert into another table after schema changes: 15297
> ```

For consistency: `s/15297/#15297/g`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15511#-:-KivYYG5Xd7XFX4zmVI-:bnfp4nl)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/schema_changer_test.go, line 1919 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15511#-KivYT2EN2vCree8yFle:-Kiv_Yx0I-V7VT_KbWBM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/554bca17f681b73cbcf8c25291c878b2b4e8d9b8/pkg/sql/schema_changer_test.go#L1919)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

For consistency: `s/15297/#15297/g`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15511)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
For release note purposes, what are the new rules for schema changes in transactions? 

---

Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15511#-:-KizaoibgmOwBhSJ-08t:bnz628a)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
We want to say that we allow schema changes with other statements in a
transaction, although some such transactions can fail:

CREATE TABLE with FK or Interleaved table, followed by statements
referencing the table https://github.com/cockroachdb/cockroach/issues/13505
DROP TABLE FOLLOWED BY CREATE fails:
https://github.com/cockroachdb/cockroach/issues/12123
database writes followed by a schema change fails: INSERT on a table,
followed by a table CREATE for example
https://github.com/cockroachdb/cockroach/issues/7570 . This is because the
transaction anchor needs to be on the system range for schema change
transactions.





On Sun, Apr 30, 2017 at 12:05 PM Ben Darnell <notifications@github.com>
wrote:

> For release note purposes, what are the new rules for schema changes in
> transactions?
> ------------------------------
>
> Review status: 0 of 6 files reviewed at latest revision, 1 unresolved
> discussion, all commit checks successful.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/15511#-:-KizaoibgmOwBhSJ-08t:bnz628a>*
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/15511#issuecomment-298240782>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALOpBKJBFOuJ-guehYDN_IYXyGZ6sAl9ks5r1LE7gaJpZM4NMHz0>
> .
>

petermattis commented:
Cc @jseldess (in case you weren't already paying attention).

jseldess commented:
Thanks, @petermattis. So to summarize, it's now safe to say that multiple schema changes, or a schema change followed by non-schema changing statements, are allowed within a single transaction, with the above exceptions?

petermattis commented:
@vivekmenezes ?

vivekmenezes commented:
Yes those three issues are the only known outstanding issues. We can't promise to fix them in 1.1
